File: Interview Problems/Parking Lot/Java Code/domain/ParkingSlot.java
package domain;

import java.util.UUID;

public class ParkingSlot {

    private final UUID id;
    private final Vehicle.VehicleType slotType;
    private final int floorNumber;
    private boolean isOccupied;

    public ParkingSlot(Vehicle.VehicleType slotType, int floorNumber) {
        this.id = UUID.randomUUID();
        this.slotType = slotType;
        this.floorNumber = floorNumber;
        this.isOccupied = false;
    }

    public void occupy() {
        if (isOccupied) throw new IllegalStateException("Slot " + id + " is already occupied");
        this.isOccupied = true;
    }

    public void release() {
        if (!isOccupied) throw new IllegalStateException("Slot " + id + " is not occupied");
        this.isOccupied = false;
    }

    public UUID getId() { return id; }
    public Vehicle.VehicleType getSlotType() { return slotType; }
    public int getFloorNumber() { return floorNumber; }
    public boolean isOccupied() { return isOccupied; }

    @Override
    public String toString() {
        return "ParkingSlot{id=" + id + ", type=" + slotType
                + ", floor=" + floorNumber + ", occupied=" + isOccupied + "}";
    }
}

File: Interview Problems/Parking Lot/Python Code/domain/parking_slot.py
import uuid
from domain.vehicle import VehicleType

class ParkingSlot:
    def __init__(self, slot_type: VehicleType, floor_number: int):
        self.id = uuid.uuid4()
        self.slot_type = slot_type
        self.floor_number = floor_number
        self.is_occupied = False

    def occupy(self):
        if self.is_occupied:
            raise RuntimeError(f"Slot {self.id} is already occupied")
        self.is_occupied = True

    def release(self):
        if not self.is_occupied:
            raise RuntimeError(f"Slot {self.id} is not occupied")
        self.is_occupied = False

    def __repr__(self):
        return f"ParkingSlot(id={self.id}, type={self.slot_type.name}, floor={self.floor_number}, occupied={self.is_occupied})"
